File: LeetCode C++/Problems/938. Range Sum of BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void Recursion(TreeNode* curr, int low, int high, int& sum) {
        if (curr == nullptr) {
            return;
        }
        // too small
        if (curr->val < low) {
            Recursion(curr->right, low, high, sum);
        } else if (curr->val > high) {
            Recursion(curr->left, low, high, sum);
        } else {
            sum += curr->val;
            Recursion(curr->left, low, high, sum);
            Recursion(curr->right, low, high, sum);
        }
    }

    int rangeSumBST(TreeNode* root, int low, int high) {
        int sum = 0;
        Recursion(root, low, high, sum);
        return sum;
    }
};

File: LeetCode Java/500-999/938. Range Sum of BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int sum;

    protected int Tilt(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int lt = this.Tilt(node.left);
        int rt = this.Tilt(node.right);
        int tilt = Math.abs(lt - rt);
        this.sum += tilt;
        return node.val + lt + rt;
    }

    public int findTilt(TreeNode root) {
        this.sum = 0;
        this.Tilt(root);
        return this.sum;
    }
}
